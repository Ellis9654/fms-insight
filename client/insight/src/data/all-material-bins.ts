/* Copyright (c) 2018, John Lenz

All rights reserved.

Redistribution and use in source and binary forms, with or without
modification, are permitted provided that the following conditions are met:

    * Redistributions of source code must retain the above copyright
      notice, this list of conditions and the following disclaimer.

    * Redistributions in binary form must reproduce the above
      copyright notice, this list of conditions and the following
      disclaimer in the documentation and/or other materials provided
      with the distribution.

    * Neither the name of John Lenz, Black Maple Software, SeedTactics,
      nor the names of other contributors may be used to endorse or
      promote products derived from this software without specific
      prior written permission.

THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
"AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR
A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT
OWNER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT
LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY
THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
(INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
 */

import * as api from "./api";
import * as im from "immutable";

export type MaterialList = ReadonlyArray<Readonly<api.IInProcessMaterial>>;

export type AllMaterialBins = im.Map<string, MaterialList>;

export function selectAllMaterialIntoBins(curSt: Readonly<api.ICurrentStatus>): AllMaterialBins {
  const palLoc = im
    .Map(curSt.pallets)
    .map(st => " (" + st.currentPalletLocation.group + " #" + st.currentPalletLocation.num.toString() + ")");

  return im
    .Seq(curSt.material)
    .map(mat => {
      switch (mat.location.type) {
        case api.LocType.InQueue:
          return { region: mat.location.currentQueue || "Queue", mat };
        case api.LocType.OnPallet:
          let region = "Pallet";
          if (mat.location.pallet) {
            region = "Pallet " + mat.location.pallet.toString() + palLoc.get(mat.location.pallet) || "";
          }
          return { region, mat };
        case api.LocType.Free:
        default:
          switch (mat.action.type) {
            case api.ActionType.Loading:
              return { region: "Raw Material", mat };
            default:
              return { region: "Free Material", mat };
          }
      }
    })
    .groupBy(x => x.region)
    .map(group =>
      group
        .map(x => x.mat)
        .sortBy(mat => {
          switch (mat.location.type) {
            case api.LocType.OnPallet:
              return mat.location.face;
            case api.LocType.InQueue:
            case api.LocType.Free:
            default:
              return mat.location.queuePosition;
          }
        })
        .valueSeq()
        .toArray()
    )
    .toMap();
}
