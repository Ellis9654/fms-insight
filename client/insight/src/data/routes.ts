/* Copyright (c) 2018, John Lenz

All rights reserved.

Redistribution and use in source and binary forms, with or without
modification, are permitted provided that the following conditions are met:

    * Redistributions of source code must retain the above copyright
      notice, this list of conditions and the following disclaimer.

    * Redistributions in binary form must reproduce the above
      copyright notice, this list of conditions and the following
      disclaimer in the documentation and/or other materials provided
      with the distribution.

    * Neither the name of John Lenz, Black Maple Software, SeedTactics,
      nor the names of other contributors may be used to endorse or
      promote products derived from this software without specific
      prior written permission.

THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
"AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR
A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT
OWNER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT
LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY
THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
(INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
 */

import { NOT_FOUND } from 'redux-first-router';
import { Seq } from 'immutable';

export enum RouteLocation {
  Dashboard = 'ROUTE_Dashboard',
  LoadMonitor = 'ROUTE_LoadMonitor',
  InspectionMonitor = 'ROUTE_Inspection',
  WashMonitor = 'ROUTE_Wash',
  CostPerPiece = 'ROUTE_CostPerPiece',
  Efficiency = 'ROUTE_Efficiency'
}

export const routeMap = {
  [RouteLocation.Dashboard]: '/',
  [RouteLocation.LoadMonitor]: '/station/loadunload/:num',
  [RouteLocation.InspectionMonitor]: '/station/inspection',
  [RouteLocation.WashMonitor]: '/station/wash',
  [RouteLocation.CostPerPiece]: '/cost',
  [RouteLocation.Efficiency]: '/efficiency',
};

export type Action =
  | { type: RouteLocation.Dashboard }
  | {
      type: RouteLocation.LoadMonitor,
      payload: { num: number },
      meta?: {
        query?: {
          queue?: ReadonlyArray<string>,
          free?: null,
        }
      },
    }
  | {
      type: RouteLocation.InspectionMonitor,
      meta?: {
        query?: {
         type?: string
        }
      },
    }
  | {
      type: RouteLocation.WashMonitor,
    }
  | { type: RouteLocation.CostPerPiece }
  | { type: RouteLocation.Efficiency }
  | { type: typeof NOT_FOUND }
  ;

export enum StationMonitorType {
  LoadUnload = 'StationType_LoadUnload',
  Inspection = 'StationType_Insp',
  Wash = 'StationType_Wash',
}

export interface State {
  readonly current: RouteLocation;
  readonly station_monitor: StationMonitorType;
  readonly selected_load_id: number;
  readonly selected_insp_type?: string;
  readonly station_queues: ReadonlyArray<string>;
  readonly station_free_material: boolean;
}

export const initial: State = {
  current: RouteLocation.Dashboard,
  station_monitor: StationMonitorType.LoadUnload,
  selected_load_id: 1,
  selected_insp_type: undefined,
  station_queues: [],
  station_free_material: false,
};

export function switchToStationMonitorPage(curSt: State): Action {
  switch (curSt.station_monitor) {
    case StationMonitorType.LoadUnload:
      return {
        type: RouteLocation.LoadMonitor,
        payload: { num: curSt.selected_load_id },
        meta: { query: { queue: curSt.station_queues, free: curSt.station_free_material ? null : undefined }},
      };

    case StationMonitorType.Inspection:
      return {
        type: RouteLocation.InspectionMonitor,
        meta: { query: {type: curSt.selected_insp_type } },
      };

    case StationMonitorType.Wash:
      return {
        type: RouteLocation.WashMonitor,
      };
  }
}

export function displayLoadStation(
    num: number, queues: ReadonlyArray<string>, freeMaterial: boolean
  ): Action {
  return {
    type: RouteLocation.LoadMonitor,
    payload: { num },
    meta: { query: {
      queue: queues.length === 0 ? undefined : queues,
      free: freeMaterial ? null : undefined,
    } }
  };
}

export function displayInspectionType(type: string | undefined): Action {
  return {
    type: RouteLocation.InspectionMonitor,
    meta: {query: { type }},
  };
}

export function displayWash(): Action {
  return {
    type: RouteLocation.WashMonitor,
  };
}

export function reducer(s: State, a: Action): State {
  if ( s === undefined) { return initial; }
  switch (a.type) {
    case RouteLocation.LoadMonitor:
      var query = (a.meta || {}).query || {};
      return {...s,
        current: RouteLocation.LoadMonitor,
        station_monitor: StationMonitorType.LoadUnload,
        selected_load_id: a.payload.num,
        station_queues: Seq(query.queue || []).take(3).toArray(),
        station_free_material: query.free === null ? true : false
      };
    case RouteLocation.InspectionMonitor:
      var iquery = (a.meta || {}).query || {};
      return {...s,
        current: RouteLocation.InspectionMonitor,
        station_monitor: StationMonitorType.Inspection,
        selected_insp_type: iquery.type,
      };
    case RouteLocation.WashMonitor:
      return {...s,
        current: RouteLocation.WashMonitor,
        station_monitor: StationMonitorType.Wash,
      };
    case RouteLocation.CostPerPiece:
      return {...s, current: RouteLocation.CostPerPiece };
    case RouteLocation.Efficiency:
      return {...s, current: RouteLocation.Efficiency };
    case RouteLocation.Dashboard:
    case NOT_FOUND:
      return {...s, current: RouteLocation.Dashboard };
    default:
      return s;
  }
}