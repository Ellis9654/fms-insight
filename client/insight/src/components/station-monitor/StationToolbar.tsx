/* Copyright (c) 2018, John Lenz

All rights reserved.

Redistribution and use in source and binary forms, with or without
modification, are permitted provided that the following conditions are met:

    * Redistributions of source code must retain the above copyright
      notice, this list of conditions and the following disclaimer.

    * Redistributions in binary form must reproduce the above
      copyright notice, this list of conditions and the following
      disclaimer in the documentation and/or other materials provided
      with the distribution.

    * Neither the name of John Lenz, Black Maple Software, SeedTactics,
      nor the names of other contributors may be used to endorse or
      promote products derived from this software without specific
      prior written permission.

THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
"AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR
A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT
OWNER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT
LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY
THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
(INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
 */
import * as React from "react";
import Select from "@material-ui/core/Select";
import MenuItem from "@material-ui/core/MenuItem";
import TextField from "@material-ui/core/TextField";
import Input from "@material-ui/core/Input";
import FormControl from "@material-ui/core/FormControl";
import Downshift from "downshift";
import Paper from "@material-ui/core/Paper";
import ArrowDropDownIcon from "@material-ui/icons/ArrowDropDown";
import List from "@material-ui/core/List";
import ListItem from "@material-ui/core/ListItem";
import ListItemText from "@material-ui/core/ListItemText";
import ListItemSecondaryAction from "@material-ui/core/ListItemSecondaryAction";
import IconButton from "@material-ui/core/IconButton";
import DeleteIcon from "@material-ui/icons/Delete";
import Typography from "@material-ui/core/Typography";
import CameraAlt from "@material-ui/icons/CameraAlt";
import SearchIcon from "@material-ui/icons/Search";
import { Tooltip } from "@material-ui/core";
import { User } from "oidc-client";
import { HashSet } from "prelude-ts";

import * as routes from "../../data/routes";
import { Store, connect, DispatchAction, mkAC } from "../../store/store";
import * as api from "../../data/api";
import * as operators from "../../data/operators";
import * as guiState from "../../data/gui-state";

interface OperatorSelectProps {
  readonly operators: HashSet<string>;
  readonly currentOperator: string | null;
  readonly setOperator: DispatchAction<operators.ActionType.SetOperator>;
  readonly removeOperator: DispatchAction<operators.ActionType.RemoveOperator>;
}

class OperatorSelect extends React.PureComponent<OperatorSelectProps> {
  render() {
    const opers = this.props.operators.toArray({ sortOn: x => x });
    return (
      <Downshift
        selectedItem={this.props.currentOperator}
        onChange={o => {
          this.props.setOperator({ operator: o });
        }}
      >
        {ds => (
          <div style={{ position: "relative" }}>
            <TextField
              InputProps={{
                // tslint:disable-next-line:no-any
                ...(ds.getInputProps({ placeholder: "Operator" }) as any),
                onKeyUp: k => {
                  if (k.keyCode === 13 && ds.inputValue && ds.inputValue.length > 0) {
                    this.props.setOperator({ operator: ds.inputValue });
                    ds.closeMenu();
                  }
                },
                endAdornment: <ArrowDropDownIcon onClick={() => ds.openMenu()} />
              }}
            />
            {ds.isOpen ? (
              <Paper
                style={{
                  position: "absolute",
                  zIndex: 1,
                  left: 0,
                  right: 0
                }}
              >
                {ds.inputValue && ds.inputValue.length > 0 && !this.props.operators.contains(ds.inputValue) ? (
                  <Typography variant="caption" align="center">
                    Press enter to add new
                  </Typography>
                ) : (
                  undefined
                )}
                <List>
                  {opers.map((o, idx) => (
                    <ListItem key={idx} button {...ds.getItemProps({ item: o })}>
                      <ListItemText primary={o} />
                      <ListItemSecondaryAction>
                        <IconButton onClick={() => this.props.removeOperator({ operator: o })}>
                          <DeleteIcon />
                        </IconButton>
                      </ListItemSecondaryAction>
                    </ListItem>
                  ))}
                </List>
              </Paper>
            ) : (
              undefined
            )}
          </div>
        )}
      </Downshift>
    );
  }
}

const toolbarStyle = {
  display: "flex",
  backgroundColor: "#E0E0E0",
  paddingLeft: "24px",
  paddingRight: "24px",
  paddingBottom: "4px",
  height: "2.5em",
  alignItems: "flex-end" as "flex-end"
};

interface StationToolbarProps {
  readonly current_route: routes.State;
  readonly queues: { [key: string]: api.IQueueSize };
  readonly insp_types: HashSet<string>;
  readonly operators: HashSet<string>;
  readonly currentOperator: string | null;
  readonly currentUser: User | null;

  readonly displayLoadStation: (num: number, queues: ReadonlyArray<string>, freeMaterial: boolean) => void;
  readonly displayInspection: (type: string | undefined) => void;
  readonly displayWash: () => void;
  readonly displayQueues: (queues: ReadonlyArray<string>, freeMaterial: boolean) => void;
  readonly displayAllMaterial: () => void;
  readonly setOperator: DispatchAction<operators.ActionType.SetOperator>;
  readonly removeOperator: DispatchAction<operators.ActionType.RemoveOperator>;
  readonly openQrCodeScan: () => void;
  readonly openManualSerial: () => void;
}

const freeMaterialSym = "@@insight_free_material@@";
const allInspSym = "@@all_inspection_display@@";

function StationToolbar(props: StationToolbarProps) {
  const queueNames = Object.keys(props.queues).sort();

  function setStation(s: string) {
    const type = s as routes.StationMonitorType;
    switch (type) {
      case routes.StationMonitorType.LoadUnload:
        props.displayLoadStation(
          props.current_route.selected_load_id,
          props.current_route.load_queues,
          props.current_route.load_free_material
        );
        break;

      case routes.StationMonitorType.Inspection:
        props.displayInspection(props.current_route.selected_insp_type);
        break;

      case routes.StationMonitorType.Wash:
        props.displayWash();
        break;

      case routes.StationMonitorType.Queues:
        props.displayQueues(props.current_route.standalone_queues, props.current_route.standalone_free_material);
        break;

      case routes.StationMonitorType.AllMaterial:
        props.displayAllMaterial();
        break;
    }
  }

  function setLoadNumber(valStr: string) {
    const val = parseFloat(valStr);
    if (!isNaN(val) && isFinite(val)) {
      props.displayLoadStation(val, props.current_route.load_queues, props.current_route.load_free_material);
    }
  }

  function setInspType(type: string) {
    props.displayInspection(type === allInspSym ? undefined : type);
  }

  let curLoadQueueCount = props.current_route.load_queues.length;
  if (curLoadQueueCount > 3) {
    curLoadQueueCount = 3;
  }

  let curStandaloneCount = props.current_route.standalone_queues.length;
  if (curStandaloneCount > 3) {
    curStandaloneCount = 3;
  }

  // the material-ui type bindings specify `e.target.value` to have type string, but
  // when multiple selects are enabled it is actually a type string[]
  // tslint:disable-next-line:no-any
  function setLoadQueues(newQueuesAny: any) {
    const newQueues = newQueuesAny as ReadonlyArray<string>;
    const free = newQueues.includes(freeMaterialSym);
    props.displayLoadStation(
      props.current_route.selected_load_id,
      newQueues.slice(0, 3).filter(q => q !== freeMaterialSym),
      free
    );
  }

  let loadqueues: string[] = [...props.current_route.load_queues];
  if (props.current_route.load_free_material) {
    loadqueues.push(freeMaterialSym);
  }

  // tslint:disable-next-line:no-any
  function setStandaloneQueues(newQueuesAny: any) {
    const newQueues = newQueuesAny as ReadonlyArray<string>;
    const free = newQueues.includes(freeMaterialSym);
    props.displayQueues(newQueues.slice(0, 3).filter(q => q !== freeMaterialSym), free);
  }

  let standalonequeues: string[] = [...props.current_route.standalone_queues];
  if (props.current_route.standalone_free_material) {
    standalonequeues.push(freeMaterialSym);
  }

  return (
    <nav style={toolbarStyle}>
      <div style={{ display: "flex", alignItems: "flex-end", flexGrow: 1 }}>
        <Select
          name="choose-station-type-select"
          value={props.current_route.station_monitor}
          onChange={e => setStation(e.target.value)}
          autoWidth
        >
          <MenuItem value={routes.StationMonitorType.LoadUnload}>Load Station</MenuItem>
          <MenuItem value={routes.StationMonitorType.Inspection}>Inspection</MenuItem>
          <MenuItem value={routes.StationMonitorType.Wash}>Wash</MenuItem>
          <MenuItem value={routes.StationMonitorType.Queues}>Queues</MenuItem>
          <MenuItem value={routes.StationMonitorType.AllMaterial}>All Material</MenuItem>
        </Select>
        {props.current_route.station_monitor === routes.StationMonitorType.LoadUnload ? (
          <Input
            type="number"
            placeholder="Load Station Number"
            key="loadnumselect"
            value={props.current_route.selected_load_id}
            onChange={e => setLoadNumber(e.target.value)}
            style={{ width: "3em", marginLeft: "1em" }}
          />
        ) : (
          undefined
        )}
        {props.current_route.station_monitor === routes.StationMonitorType.Inspection ? (
          <Select
            key="inspselect"
            value={props.current_route.selected_insp_type || allInspSym}
            onChange={e => setInspType(e.target.value)}
            style={{ marginLeft: "1em" }}
          >
            <MenuItem key={allInspSym} value={allInspSym}>
              <em>All</em>
            </MenuItem>
            {props.insp_types.toArray({ sortOn: x => x }).map(ty => (
              <MenuItem key={ty} value={ty}>
                {ty}
              </MenuItem>
            ))}
          </Select>
        ) : (
          undefined
        )}
        {props.current_route.station_monitor === routes.StationMonitorType.LoadUnload ? (
          <FormControl style={{ marginLeft: "1em" }}>
            {loadqueues.length === 0 ? (
              <label
                style={{
                  position: "absolute",
                  top: "24px",
                  left: 0,
                  color: "rgba(0,0,0,0.54)",
                  fontSize: "0.9rem"
                }}
              >
                Display queue(s)
              </label>
            ) : (
              undefined
            )}
            <Select
              multiple
              name="station-monitor-queue-select"
              data-testid="station-monitor-queue-select"
              key="queueselect"
              displayEmpty
              value={loadqueues}
              inputProps={{ id: "queueselect" }}
              style={{ minWidth: "10em" }}
              onChange={e => setLoadQueues(e.target.value)}
            >
              <MenuItem key={freeMaterialSym} value={freeMaterialSym}>
                Free Material
              </MenuItem>
              {queueNames.map((q, idx) => (
                <MenuItem key={idx} value={q}>
                  {q}
                </MenuItem>
              ))}
            </Select>
          </FormControl>
        ) : (
          undefined
        )}
        {props.current_route.station_monitor === routes.StationMonitorType.Queues ? (
          <FormControl style={{ marginLeft: "1em", minWidth: "10em" }}>
            {standalonequeues.length === 0 ? (
              <label
                style={{
                  position: "absolute",
                  top: "24px",
                  left: 0,
                  color: "rgba(0,0,0,0.54)",
                  fontSize: "0.9rem"
                }}
              >
                Select queue(s)
              </label>
            ) : (
              undefined
            )}
            <Select
              multiple
              name="station-monitor-queue-select"
              data-testid="station-monitor-queue-select"
              key="queueselect"
              displayEmpty
              value={standalonequeues}
              inputProps={{ id: "queueselect" }}
              onChange={e => setStandaloneQueues(e.target.value)}
            >
              <MenuItem key={freeMaterialSym} value={freeMaterialSym}>
                Free Material
              </MenuItem>
              {queueNames.map((q, idx) => (
                <MenuItem key={idx} value={q}>
                  {q}
                </MenuItem>
              ))}
            </Select>
          </FormControl>
        ) : (
          undefined
        )}
        {window.location.protocol === "https:" || window.location.hostname === "localhost" ? (
          <Tooltip title="Scan QR Code">
            <IconButton onClick={props.openQrCodeScan} style={{ height: "25px", paddingTop: 0, paddingBottom: 0 }}>
              <CameraAlt />
            </IconButton>
          </Tooltip>
        ) : (
          undefined
        )}
        <Tooltip title="Enter Serial">
          <IconButton onClick={props.openManualSerial} style={{ height: "25px", paddingTop: 0, paddingBottom: 0 }}>
            <SearchIcon />
          </IconButton>
        </Tooltip>
      </div>
      <div>
        {props.currentUser ? (
          <div>{props.currentUser.profile.name || props.currentUser.profile.sub}</div>
        ) : (
          <OperatorSelect
            operators={props.operators}
            currentOperator={props.currentOperator}
            setOperator={props.setOperator}
            removeOperator={props.removeOperator}
          />
        )}
      </div>
    </nav>
  );
}

export default connect(
  (st: Store) => ({
    current_route: st.Route,
    queues: st.Current.current_status.queues,
    insp_types: st.Events.last30.mat_summary.inspTypes,
    operators: st.Operators.operators,
    currentOperator: st.Operators.current || null,
    currentUser: st.ServerSettings.user || null
  }),
  {
    displayLoadStation: routes.displayLoadStation,
    displayInspection: routes.displayInspectionType,
    displayWash: routes.displayWash,
    displayQueues: routes.displayQueues,
    displayAllMaterial: routes.displayAllMaterial,
    setOperator: mkAC(operators.ActionType.SetOperator),
    removeOperator: mkAC(operators.ActionType.RemoveOperator),
    openQrCodeScan: () => ({
      type: guiState.ActionType.SetScanQrCodeDialog,
      open: true
    }),
    openManualSerial: () => ({
      type: guiState.ActionType.SetManualSerialEntryDialog,
      open: true
    })
  }
)(StationToolbar);
