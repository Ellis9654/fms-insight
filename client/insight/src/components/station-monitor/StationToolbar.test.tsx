/* Copyright (c) 2018, John Lenz

All rights reserved.

Redistribution and use in source and binary forms, with or without
modification, are permitted provided that the following conditions are met:

    * Redistributions of source code must retain the above copyright
      notice, this list of conditions and the following disclaimer.

    * Redistributions in binary form must reproduce the above
      copyright notice, this list of conditions and the following
      disclaimer in the documentation and/or other materials provided
      with the distribution.

    * Neither the name of John Lenz, Black Maple Software, SeedTactics,
      nor the names of other contributors may be used to endorse or
      promote products derived from this software without specific
      prior written permission.

THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
"AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR
A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT
OWNER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT
LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY
THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
(INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
 */

import * as React from 'react';
import { shallow } from 'enzyme';

import { StationToolbar } from './StationToolbar';
import * as routes from '../../data/routes';

const basicState: routes.State = {
  current: routes.RouteLocation.StationMonitor,
  selected_station_type: routes.SelectedStationType.Inspection,
  selected_station_id: 4,
  station_queues: ["a"],
};

it('displays the toolbar for inspection with one queue', () => {
  const setRoute = jest.fn();

  let val = shallow(
    <StationToolbar
      current_route={basicState}
      setStationRoute={setRoute}
    />);
  expect(val).toMatchSnapshot('inspection toolbar');
});

it('displays the toolbar for wash with no queues', () => {
  const setRoute = jest.fn();
  const st = {...basicState,
    selected_station_type: routes.SelectedStationType.Wash,
    station_queues: [],
  };

  let val = shallow(
    <StationToolbar
      current_route={st}
      setStationRoute={setRoute}
    />);
  expect(val).toMatchSnapshot('wash toolbar');
});

it('displays the toolbar for load with three queues', () => {
  const setRoute = jest.fn();
  const st = {...basicState,
    selected_station_type: routes.SelectedStationType.Inspection,
    station_queues: ["a", "b", "c"],
  };

  let val = shallow(
    <StationToolbar
      current_route={st}
      setStationRoute={setRoute}
    />);
  expect(val).toMatchSnapshot('load toolbar');
});

it("changes the station", () => {
  const setRoute = jest.fn();

  const val = shallow(
    <StationToolbar
      current_route={basicState}
      setStationRoute={setRoute}
    />);

  // tslint:disable-next-line:no-any
  const onChange = val.find("WithStyles(Select)").first().prop("onChange") as any;
  onChange({target: {value: routes.SelectedStationType.Wash}});

  expect(setRoute).toHaveBeenCalledWith(
    routes.SelectedStationType.Wash,
    basicState.selected_station_id,
    basicState.station_queues
  );
});

it("changes the station number", () => {
  const setRoute = jest.fn();

  const val = shallow(
    <StationToolbar
      current_route={basicState}
      setStationRoute={setRoute}
    />);

  // tslint:disable-next-line:no-any
  const onChange = val.find("WithStyles(Input)").first().prop("onChange") as any;
  onChange({target: {value: "12"}});

  expect(setRoute).toHaveBeenCalledWith(
    basicState.selected_station_type,
    12,
    basicState.station_queues
  );
});

it("changes the queues", () => {
  const setRoute = jest.fn();

  const val = shallow(
    <StationToolbar
      current_route={basicState}
      setStationRoute={setRoute}
    />);

  // tslint:disable-next-line:no-any
  const onChange = val.find("WithStyles(Select)").last().prop("onChange") as any;

  onChange({target: {value: "4"}});
  expect(setRoute).toHaveBeenCalledWith(
    basicState.selected_station_type,
    basicState.selected_station_id,
    ["a", "", "", ""]
  );

  setRoute.mockReset();

  onChange({target: {value: "0"}});
  expect(setRoute).toHaveBeenCalledWith(
    basicState.selected_station_type,
    basicState.selected_station_id,
    []
  );
});