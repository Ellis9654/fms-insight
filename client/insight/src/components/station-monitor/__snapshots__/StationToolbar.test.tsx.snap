// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`inspection toolbar 1`] = `
<nav
  style={
    Object {
      "alignItems": "center",
      "backgroundColor": "#E0E0E0",
      "display": "flex",
      "height": "2.5em",
      "justifyContent": "space-evenly",
      "paddingLeft": "24px",
      "paddingRight": "24px",
    }
  }
>
  <div>
    <WithStyles(Select)
      autoWidth={true}
      onChange={[Function]}
      value="inspection"
    >
      <WithStyles(MenuItem)
        value="loadunload"
      >
        Load Station
      </WithStyles(MenuItem)>
      <WithStyles(MenuItem)
        value="inspection"
      >
        Inspection
      </WithStyles(MenuItem)>
      <WithStyles(MenuItem)
        value="wash"
      >
        Wash
      </WithStyles(MenuItem)>
    </WithStyles(Select)>
    <WithStyles(Input)
      onChange={[Function]}
      style={
        Object {
          "marginLeft": "1em",
          "width": "3em",
        }
      }
      type="number"
      value={4}
    />
  </div>
  <div>
    <WithStyles(Select)
      autoWidth={true}
      onChange={[Function]}
      value={1}
    >
      <WithStyles(MenuItem)
        value={0}
      >
        No Queues
      </WithStyles(MenuItem)>
      <WithStyles(MenuItem)
        value={1}
      >
        One Queue
      </WithStyles(MenuItem)>
      <WithStyles(MenuItem)
        value={2}
      >
        Two Queues
      </WithStyles(MenuItem)>
      <WithStyles(MenuItem)
        value={4}
      >
        Four Queues
      </WithStyles(MenuItem)>
    </WithStyles(Select)>
  </div>
</nav>
`;

exports[`load toolbar 1`] = `
<nav
  style={
    Object {
      "alignItems": "center",
      "backgroundColor": "#E0E0E0",
      "display": "flex",
      "height": "2.5em",
      "justifyContent": "space-evenly",
      "paddingLeft": "24px",
      "paddingRight": "24px",
    }
  }
>
  <div>
    <WithStyles(Select)
      autoWidth={true}
      onChange={[Function]}
      value="inspection"
    >
      <WithStyles(MenuItem)
        value="loadunload"
      >
        Load Station
      </WithStyles(MenuItem)>
      <WithStyles(MenuItem)
        value="inspection"
      >
        Inspection
      </WithStyles(MenuItem)>
      <WithStyles(MenuItem)
        value="wash"
      >
        Wash
      </WithStyles(MenuItem)>
    </WithStyles(Select)>
    <WithStyles(Input)
      onChange={[Function]}
      style={
        Object {
          "marginLeft": "1em",
          "width": "3em",
        }
      }
      type="number"
      value={4}
    />
  </div>
  <div>
    <WithStyles(Select)
      autoWidth={true}
      onChange={[Function]}
      value={4}
    >
      <WithStyles(MenuItem)
        value={0}
      >
        No Queues
      </WithStyles(MenuItem)>
      <WithStyles(MenuItem)
        value={1}
      >
        One Queue
      </WithStyles(MenuItem)>
      <WithStyles(MenuItem)
        value={2}
      >
        Two Queues
      </WithStyles(MenuItem)>
      <WithStyles(MenuItem)
        value={4}
      >
        Four Queues
      </WithStyles(MenuItem)>
    </WithStyles(Select)>
  </div>
</nav>
`;

exports[`wash toolbar 1`] = `
<nav
  style={
    Object {
      "alignItems": "center",
      "backgroundColor": "#E0E0E0",
      "display": "flex",
      "height": "2.5em",
      "justifyContent": "space-evenly",
      "paddingLeft": "24px",
      "paddingRight": "24px",
    }
  }
>
  <div>
    <WithStyles(Select)
      autoWidth={true}
      onChange={[Function]}
      value="wash"
    >
      <WithStyles(MenuItem)
        value="loadunload"
      >
        Load Station
      </WithStyles(MenuItem)>
      <WithStyles(MenuItem)
        value="inspection"
      >
        Inspection
      </WithStyles(MenuItem)>
      <WithStyles(MenuItem)
        value="wash"
      >
        Wash
      </WithStyles(MenuItem)>
    </WithStyles(Select)>
    <WithStyles(Input)
      onChange={[Function]}
      style={
        Object {
          "marginLeft": "1em",
          "width": "3em",
        }
      }
      type="number"
      value={4}
    />
  </div>
  <div>
    <WithStyles(Select)
      autoWidth={true}
      onChange={[Function]}
      value={0}
    >
      <WithStyles(MenuItem)
        value={0}
      >
        No Queues
      </WithStyles(MenuItem)>
      <WithStyles(MenuItem)
        value={1}
      >
        One Queue
      </WithStyles(MenuItem)>
      <WithStyles(MenuItem)
        value={2}
      >
        Two Queues
      </WithStyles(MenuItem)>
      <WithStyles(MenuItem)
        value={4}
      >
        Four Queues
      </WithStyles(MenuItem)>
    </WithStyles(Select)>
  </div>
</nav>
`;
